internal/locate/region_cache.go

```diff
@@ -1602,6 +1636,46 @@ func (c *RegionCache) scanRegionsFromCache(bo *retry.Backoffer, startKey, endKey
 	return regions, nil
 }
 
+func (c *RegionCache) timelyRefreshCache(intervalS uint64) {
+	if intervalS <= 0 {
+		return
+	}
+	ticker := time.NewTicker(time.Duration(intervalS) * time.Second)
+	go func() {
+		defer ticker.Stop()
+		for {
+			select {
+			case <-c.ctx.Done():
+				return
+			case <-ticker.C:
+				intervalMs := int(1000 * intervalS)
+				if err := c.refreshRegionIndex(retry.NewBackofferWithVars(c.ctx, intervalMs, nil)); err != nil {
+					logutil.BgLogger().Error("refresh region cache failed", zap.Error(err))
+				}
+			}
+		}
+	}()
+}
+
+func (c *RegionCache) refreshRegionIndex(bo *retry.Backoffer) error {
+	totalRegions := make([]*Region, 0)
+	startKey := []byte{}
+	for {
+		regions, err := c.scanRegions(bo, startKey, nil, 10000)
+		if err != nil {
+			return err
+		}
+		if len(regions) > 0 && regions[len(regions)-1].meta.EndKey == nil {
+			totalRegions = append(totalRegions, regions...)
+			break
+		}
+		totalRegions = append(totalRegions, regions...)
+		startKey = regions[len(regions)-1].meta.EndKey
```

Comment:
it should be better like this:

totalRegions = append(totalRegions, regions...)
if len(regions) == 0 || len(regions[len(regions)-1].meta.EndKey) == 0 {
   break
}
startKey = regions[len(regions)-1].meta.EndKey

Reply:
updated